src/plugins/ldif/ldif.c

```diff
@@ -0,0 +1,331 @@
+			       keyNew ("system:/elektra/modules/ldif/exports/error", KEY_FUNC, elektraLdifError, KEY_END),
+			       keyNew ("system:/elektra/modules/ldif/exports/checkconf", KEY_FUNC, elektraLdifCheckConf, KEY_END),
+#include ELEKTRA_README
+			       keyNew ("system:/elektra/modules/ldif/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+	// get all keys
+	int errnosave = errno;
+	const char * filename = keyString (parentKey);
+	LDIFFP * lfp = ldif_open(filename, "r");
+
+	if (!lfp)
+	{
+		ELEKTRA_SET_ERROR_GET (parentKey);
+		errno = errnosave;
+		return ELEKTRA_PLUGIN_STATUS_ERROR;
+	}
+
+	unsigned long lineno = 0;
+	char *buff = NULL;
+	int buflen = 0;
+
+	ssize_t ksize = 0;
+
+	// TODO replace printf with ELEKTRA_LOG_DEBUG
+	while (ldif_read_record (lfp, &lineno, &buff, &buflen))
+	{
+		char * line;
+		char * next = buff;
+
+		while ((line = ldif_getline (&next)) != NULL)
+		{
+			char * type = NULL;
+			char * value = NULL;
+			ber_len_t vlen;
+
+			if (ldif_parse_line (line, &type, &value, &vlen))
+			{
+				ELEKTRA_ADD_VALIDATION_SYNTACTIC_WARNINGF (parentKey, "cannot parse line: %s\n", line);
+				continue;
+			}
+
+			if (elektraStrCmp (type, "dn") == 0)
+			{
+				// key
+				printf ("found key: %s\n", value);
+
+				char ** tokens = parseToken (value);
+
+				if (tokens == NULL)
+				{
+					ELEKTRA_SET_ERROR_GET (parentKey);
+
+					elektraFree (buff);
+					ldif_close (lfp);
+					return ELEKTRA_PLUGIN_STATUS_ERROR;
+				}
+
+				int size = 0;
+
+				while (tokens[size] != NULL)
+				{
+					size++;
+				}
+
+				reverse (tokens, size);
+
+				for (int i = 0; i < size; ++i)
+				{
+
+					Key * key = keyNew (keyName (parentKey), KEY_END);
+					char * domainpart = makeKey (tokens, i + 1);
+
+					printf ("saving key: %s\n", domainpart);
+
+					if (domainpart == NULL)
+					{
+						ELEKTRA_SET_ERROR_GET (parentKey);
+						elektraFree (domainpart);
+						elektraFree (tokens);
+						elektraFree (buff);
+						ldif_close (lfp);
+						return ELEKTRA_PLUGIN_STATUS_ERROR;
+					}
+
+					if (keyAddName (key, domainpart) == -1)
+					{
+						ELEKTRA_ADD_VALIDATION_SYNTACTIC_WARNINGF (
+							parentKey, "Key name '%s' is not valid, discarding key", value);
+						elektraFree (domainpart);
+						keyDel (key);
+						break;
+					}
+
+					if (ksAppendKey (returned, key) != ksize + 1)
+					{
+						// TODO re-enable duplicate check
+//						ELEKTRA_SET_VALIDATION_SYNTACTIC_ERRORF (parentKey,
+//											 "Duplicated key '%s' at position %ld in file %s",
+//											 keyName (key), ftell (lfp->fp), filename);
+//						free (buff);
+//						ldif_close (lfp);
+//						return ELEKTRA_PLUGIN_STATUS_ERROR;
+					}
+					++ksize;
+					elektraFree (domainpart);
+				}
+
+				elektraFree (tokens);
+			}
+			else
+			{
+				// value
+				printf ("found value: type: %s, value: %s\n", type, value);
```

Comment:
please use logger

Reply:
This should be done in a6a767dfc5b9a5fb9b8387fe75864bdb1ac77db6 is this correct or did I miss something here since I could not find any further calls to this function? (Beside always setting the lineno to 0)

Reply:
The `elektraLog` function is designed to be called via `ELEKTRA_LOG_DEBUG` (and similar macros), which automatically sets `filename` and `lineno` correctly.

Reply:
See also https://www.libelektra.org/tutorials/logging

Reply:
ok thank you. this should be fixed now